### PR TITLE
Add metrics for requests rejected due to per-instance limits

### PR DESCRIPTION
**Pull request by 56quarters**
#### What this PR does

Add metrics to distributors and ingesters that are incremented when we reject requests due to hitting per-instance limits.

#### Which issue(s) this PR fixes or relates to

Related #5494 #397

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -334,6 +339,12 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 		discardedExemplarsRateLimited:     validation.DiscardedExemplarsCounter(reg, validation.ReasonRateLimited),
 		discardedMetadataRateLimited:      validation.DiscardedMetadataCounter(reg, validation.ReasonRateLimited),
 
+		discardedInstanceRequests: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "cortex",
+			Name:      "distributor_instance_discarded_requests_total",
+			Help:      "Requests discarded for hitting per-instance limits",
+		}, []string{"reason"}),
+
```
Comment (charleskorn) [inline pkg/distributor/distributor.go:347]: Should we initialise the counter to 0 for each `reason` so that we can show it in dashboards even if no requests have been discarded?

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -1013,11 +1024,13 @@ func (d *Distributor) limitsMiddleware(next push.Func) push.Func {
 
 		il := d.getInstanceLimits()
 		if il.MaxInflightPushRequests > 0 && inflight > int64(il.MaxInflightPushRequests) {
+			d.discardedInstanceRequests.WithLabelValues(validation.ReasonDistributorMaxInflightPushRequests).Inc()
 			return nil, errMaxInflightRequestsReached
 		}
 
 		if il.MaxIngestionRate > 0 {
 			if rate := d.ingestionRate.Rate(); rate >= il.MaxIngestionRate {
+				d.discardedInstanceRequests.WithLabelValues(validation.ReasonDistributorMaxIngestionRate).Inc()
```
Comment (charleskorn) [inline pkg/distributor/distributor.go:1033]: Should we add / extend a test to verify that the counter is incremented when these errors are hit? 

(also applies for the max-inflight push request bytes limit below and the ingester limits)

File: pkg/ingester/metrics.go [hunk 3]
```diff
@@ -60,7 +60,8 @@ type ingesterMetrics struct {
 	// Open all existing TSDBs metrics
 	openExistingTSDB prometheus.Counter
 
-	discarded *discardedMetrics
+	discarded         *discardedMetrics
+	discardedInstance *prometheus.CounterVec
```
Comment (charleskorn) [inline pkg/ingester/metrics.go:64]: Should we add this to `discardedMetrics`? It's not clear to me what belongs in there and what goes here (we've got the discarded metadata metrics below, but everything else related to discarding data seems to be on `discardedMetrics`)

File: CHANGELOG.md [hunk 4]
```diff
@@ -55,6 +55,7 @@
 * [ENHANCEMENT] Distributor: add support for ingesting exponential histograms that are over the native histogram scale limit of 8 in OpenTelemetry format by downscaling them. #5532
 * [ENHANCEMENT] General: buffered logging: #5506
   * `-log.buffered`: Enable buffered logging
+* [ENHANCEMENT] Distributor Ingester: Add metrics to count the number of requests rejected for hitting per-instance limits, `cortex_distributor_instance_discarded_requests_total` and `cortex_ingester_instance_discarded_requests_total` respectively. #5551
```
Comment (pstibrany) [inline CHANGELOG.md:58]: We use "discarded" in other context already -- requests for which we return 4xx errors due to various limits.

These requests are rejected with 5xx error instead, and we expect that they will be repeated by client. I'd suggest using different term than "discarded" (eg. rejected, as you use in comments)

File: pkg/distributor/distributor.go [hunk 5]
```diff
@@ -324,8 +328,9 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 			Help:      "The configured replication factor.",
 		}),
 		latestSeenSampleTimestampPerUser: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
-			Name: "cortex_distributor_latest_seen_sample_timestamp_seconds",
-			Help: "Unix timestamp of latest received sample per user.",
+			Namespace: "cortex",
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:331]: This is unnecessary change in this PR, and imho also a change in wrong direction. It makes finding metrics more difficult (instead of searching for full metric name like `cortex_distributor_latest_seen_sample_timestamp_seconds`, one needs to remove prefix and try again). I'd suggest 1) not do it here, and 2) in other PR remove all usage of `Namespace` field altogether.

File: pkg/distributor/distributor.go [hunk 6]
```diff
@@ -1013,11 +1024,13 @@ func (d *Distributor) limitsMiddleware(next push.Func) push.Func {
 
 		il := d.getInstanceLimits()
 		if il.MaxInflightPushRequests > 0 && inflight > int64(il.MaxInflightPushRequests) {
+			d.discardedInstanceRequests.WithLabelValues(validation.ReasonDistributorMaxInflightPushRequests).Inc()
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:1027]: Similar to my previous comment on "discard", these are not "validation" errors. Validation errors indicate problem in the request / tenant limit.

Instead of introducing new "reasons" in `validation` package, I'd suggest to use something like

```suggestion
			d.discardedInstanceRequests.WithLabelValues(globalerror.DistributorMaxInflightPushRequests.String()).Inc()
```

(This requires new `String()`  method)

File: pkg/distributor/distributor.go (hunk 5 above)
Comment (56quarters) [reply pkg/distributor/distributor.go:331]: Fiiiiiine

File: pkg/distributor/distributor.go (hunk 6 above)
Comment (56quarters) [reply pkg/distributor/distributor.go:1027]: I can move it but I put them in `validation` because that's where the label-form for other `globalerror` errors live. I wanted to keep all the labels for error metrics in the same place.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (56quarters) [reply pkg/distributor/distributor.go:1033]: I took a look at doing this. Some of them are racey to try to recreate in a test (ingestion rate, in-flight requests). I'd rather not test these than add something that has the potential to flake. Maybe we can revisit this in the future if we run into issues where we broke these metrics without noticing.

File: pkg/ingester/metrics.go (hunk 3 above)
Comment (56quarters) [reply pkg/ingester/metrics.go]: Changed to `rejected` per Peter's comments so they're now separate and named differently.

File: pkg/distributor/distributor.go (hunk 6 above)
Comment (56quarters) [reply pkg/distributor/distributor.go]: Added a `.LabelValue()` to `globalerror.ID` which is now used everywhere.

File: CHANGELOG.md (hunk 4 above)
Comment (56quarters) [reply CHANGELOG.md:58]: Renamed.

**Review by pstibrany — APPROVED** [review]
Thank you for addressing my feedback.
